airflow/dag_processing/collection.py

```diff
@@ -547,3 +547,19 @@ def add_asset_trigger_references(
     def _encrypt_trigger_kwargs(trigger: BaseTrigger) -> tuple[str, str]:
         classpath, kwargs = trigger.serialize()
         return classpath, Trigger.encrypt_kwargs(kwargs)
+
+    @staticmethod
+    def _get_trigger_hash(classpath: str, kwargs: dict[str, Any]) -> int:
+        """
+        Return the hash of the trigger classpath and kwargs. This is used to uniquely identify a trigger.
+
+        We do not want to move this logic in a `__hash__` method in `BaseTrigger` because we do not want to
+        make the triggers hashable. The reason being, when the triggerer retrieve the list of triggers, we do
+        not want it dedupe them. When used to defer tasks, 2 triggers can have the same classpath and kwargs.
+        This is not true for event driven scheduling.
+        """
+        return hash((classpath, frozenset(kwargs.items())))
```

Comment:
This implies all arguments passed to the trigger must be hashable. Is it a known requirement?

Reply:
```
def serialize(self) -> tuple[str, dict[str, Any]]:
```

Nop 😢 Good catch. I'll need to find something else then

Reply:
`frozenset` is always hashable, no?
```python
>>> kwargs = {"arg1": 1, "arg2": "value", "arg3": set({1, 2, 3}), "arg4": {"k": "v", "kk": 1}, "arg5": b"x"}
>>> hash(kwargs)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
>>> hash(frozenset(kwargs))
-1961075596996639406
```
🤔

Reply:
A container is only hashable if all its content is. This fails say if `kwargs` is `{"dag_ids": {"a", "b", "c"}}`.

Reply:
My example already contains a set and it works without any problem.

Same for your example:
```python
>>> kwargs = {"dag_ids": {"a", "b", "c"}}
>>> hash(("<class_path>", frozenset(kwargs)))
1524053278580373800
```

> A container is only hashable if all its content is.

That's true, but as far as I know, `frozenset` is always hashable for any provided iterable.

Reply:
Actually you're right, this fails:
```
>>> hash(("<class_path>", frozenset(kwargs.items())))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'set'
```
In my first example, it uses the keys of the dict and not the values 🤦‍♂️

Reply:
since we're sure that `repr` works for all arguments, what about using:
```python
hash((classpath, frozenset(repr(kwargs.items()))))
```

Reply:
the problem is that repr is not a well-defined thing for all things that could be passed to kwargs.  so you could get unexpected results.  feels like we're missing something to clearly identify and distinguish these trigger objects in this context.

Reply:
Unfortunately, `classpath` and `kwargs` is the only information available to compare 2 triggers. I do not have other choice than using these 2 pieces of information. One solution could be to implement a solution which "tries" to handle all types like:

```
def hash_dict(obj: dict) -> int:    
    def hash_any(value):
        """Handles hashing for any type of object."""
        if isinstance(value, dict):
            return hash_dict(value)  # Recursively hash nested dictionaries
        elif isinstance(value, list):
            return hash(tuple(hash_any(item) for item in value))  # Hash list as tuple of hashes
        elif isinstance(value, set):
            return hash(frozenset(hash_any(item) for item in value))  # Hash set as frozenset
        elif isinstance(value, tuple):
            return hash(tuple(hash_any(item) for item in value))  # Handle nested tuples
        elif isinstance(value, (int, float, str, bool, type(None))):  # Simple hashable types
            return hash(value)
        else:
            raise ValueError(f"Unsupported type for hashing: {type(value)}")
    
    # Create a sorted tuple of key-value pairs (keys and values are recursively hashed)
    sorted_items = tuple(sorted((hash_any(k), hash_any(v)) for k, v in obj.items()))
    return hash(sorted_items)
```

Reply:
Or another option would to convert `kwargs` type from `dict[str, Any]` to `dict[str, Hashable]`. However, that'd be a breaking change that could break a lot of custom triggers

Reply:
After discussion with @dstandish, we decided (and thanks again for pointing out the solution Dan) to serialize `kwargs` (this is already done because we are storing these kwargs in the DB) and then hash it

Reply:
PR updated